src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DynamicInterfaceCastableImplementation.cs

```diff
@@ -0,0 +1,133 @@
+                new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.InterfaceMethodsMissingImplementationDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+                isPortedFxCopRule: false,
+                isDataflowRule: false);
+
+        internal const string MethodsDeclaredOnImplementationTypeMustBeSealedRuleId = "CA2254";
+
+        private static readonly DiagnosticDescriptor MethodsDeclaredOnImplementationTypeMustBeSealed =
+            DiagnosticDescriptorHelper.Create(
+                MethodsDeclaredOnImplementationTypeMustBeSealedRuleId,
+                new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.MethodsDeclaredOnImplementationTypeMustBeSealedTitle), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+                new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.MethodsDeclaredOnImplementationTypeMustBeSealedMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+                DiagnosticCategory.Usage,
+                RuleLevel.BuildWarning,
+                new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.MethodsDeclaredOnImplementationTypeMustBeSealedDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+                isPortedFxCopRule: false,
+                isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            DynamicInterfaceCastableImplementationUnsupported,
+            InterfaceMethodsMissingImplementation,
+            MethodsDeclaredOnImplementationTypeMustBeSealed);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterSymbolAction(context => AnalyzeType(context), SymbolKind.NamedType);
+        }
+
+        private const string DynamicInterfaceCastableImplementationAttributeTypeName = "System.Runtime.InteropServices.DynamicInterfaceCastableImplementationAttribute";
+
+        private static void AnalyzeType(SymbolAnalysisContext context)
+        {
+            INamedTypeSymbol targetType = (INamedTypeSymbol)context.Symbol;
+
+            if (targetType.TypeKind != TypeKind.Interface)
+            {
+                return;
+            }
+
+            bool isDynamicInterfaceImplementation = false;
+            foreach (var attribute in targetType.GetAttributes())
+            {
+                if (attribute.AttributeClass.ToDisplayString(SymbolDisplayFormats.QualifiedTypeAndNamespaceSymbolDisplayFormat) == DynamicInterfaceCastableImplementationAttributeTypeName)
```

Comment:
I think it's better to retrieve the attribute type in compilation start action, bail-out early if it doesn't exist, and perform symbol equality here (using the HasAttribute extension):

https://github.com/dotnet/roslyn-analyzers/blob/0f20ec1f3877f114800347a873c86d21f377f01b/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L595